dex/pair/src/safe_price_view.rs

```diff
@@ -517,69 +527,104 @@ pub trait SafePriceViewModule:
         } else {
-            left_index = current_index;
+            left_index = current_index + 1;
             right_index = price_observations.len();
         }
 
-        let mut closest_observation = observation_at_index_1.clone();
-        let mut min_timestamp_diff =
-            if target_timestamp > observation_at_index_1.recording_timestamp {
-                target_timestamp - observation_at_index_1.recording_timestamp
-            } else {
-                observation_at_index_1.recording_timestamp - target_timestamp
-            };
-
         while left_index <= right_index {
             search_index = (left_index + right_index) / 2;
-            let current_observation = price_observations.get(search_index);
-            let current_timestamp_diff =
-                if target_timestamp > current_observation.recording_timestamp {
-                    target_timestamp - current_observation.recording_timestamp
-                } else {
-                    current_observation.recording_timestamp - target_timestamp
-                };
-
-            if current_timestamp_diff < min_timestamp_diff {
-                min_timestamp_diff = current_timestamp_diff;
-                closest_observation = current_observation.clone();
-            }
-
-            match current_observation
-                .recording_timestamp
-                .cmp(&target_timestamp)
-            {
-                Ordering::Equal => return current_observation,
+            let price_observation = price_observations.get(search_index);
+            match price_observation.recording_timestamp.cmp(&target_timestamp) {
+                Ordering::Equal => return (price_observation, search_index),
                 Ordering::Less => left_index = search_index + 1,
                 Ordering::Greater => right_index = search_index - 1,
             }
         }
 
-        closest_observation
+        (PriceObservation::default(), search_index)
+    }
+
+    fn interpolate_round_by_timestamp(
+        &self,
+        price_observations: &VecMapper<PriceObservation<Self::Api>, ManagedAddress>,
+        target_timestamp: Timestamp,
+        search_index: usize,
+    ) -> Round {
+        let last_found_observation = price_observations.get(search_index);
+        let left_observation;
+        let right_observation;
+
+        if last_found_observation.recording_timestamp < target_timestamp {
+            left_observation = last_found_observation;
+            let right_observation_index = (search_index % MAX_OBSERVATIONS) + 1;
+            right_observation = price_observations.get(right_observation_index);
+        } else {
+            let left_observation_index = if search_index == 1 {
+                MAX_OBSERVATIONS
+            } else {
+                search_index - 1
+            };
+            left_observation = price_observations.get(left_observation_index);
+            right_observation = last_found_observation;
+        };
```

Comment:
Not a fan of `let` bindings without assigning anything to them if it's not strictly needed by the code. Might need some formatting after though.

```suggestion
        let (left_observation, right_observation) = if last_found_observation.recording_timestamp < target_timestamp {
            let right_observation_index = (search_index % MAX_OBSERVATIONS) + 1;
            (last_found_observation, price_observations.get(right_observation_index))
        } else {
            let left_observation_index = if search_index == 1 {
                MAX_OBSERVATIONS
            } else {
                search_index - 1
            };
            (price_observations.get(left_observation_index), last_found_observation)
        };
```

Reply:
Updated.